addons/stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -17,48 +17,121 @@ class StockPickingBatch(models.Model):
+            allowed_picking_ids = [picking['id'] for picking in allowed_pickings]
+            batch.allowed_picking_ids = [(6, False, allowed_picking_ids)]
+
+    @api.depends('picking_ids', 'picking_ids.move_lines')
+    def _compute_move_ids(self):
+        for batch in self:
+            batch.move_ids = batch.picking_ids.move_lines
+            batch.show_check_availability = any(m.state != 'assigned' for m in batch.move_ids)
+
+
+    # -------------------------------------------------------------------------
+    # CRUD
+    # -------------------------------------------------------------------------
     @api.model
     def create(self, vals):
         if vals.get('name', '/') == '/':
             vals['name'] = self.env['ir.sequence'].next_by_code('picking.batch') or '/'
-        return super(StockPickingBatch, self).create(vals)
+        res = super().create(vals)
+        res._sanity_check()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('picking_type_id'):
+            self._sanity_check()
+        if vals.get('picking_ids'):
+            batch_without_picking_type = self.filtered(lambda batch: not batch.picking_type_id)
+            if batch_without_picking_type:
+                picking_command = vals.get('picking_ids')[0]
+                picking_id = picking_command[2] and picking_command[2][0]
+                picking = self.env['stock.picking'].browse(picking_id)
+                batch_without_picking_type.picking_type_id = picking.picking_type_id.id
+                self._sanity_check()
+        return res
+
+    def unlink(self):
+        if any(batch.state != 'draft' for batch in self):
+            raise UserError(_("You can only delete draft batch transfers."))
+        return super().unlink()
 
-    def confirm_picking(self):
+    # -------------------------------------------------------------------------
+    # Action methods
+    # -------------------------------------------------------------------------
+    def action_confirm(self):
+        """Sanity checks, assign the pickings and mark the batch as confirmed."""
+        self.ensure_one()
+        if not self.picking_ids:
+            raise UserError(_("You have to set some pickings to batch."))
+        self.picking_ids.action_assign()
         self._check_company()
-        pickings_todo = self.mapped('picking_ids')
-        self.write({'state': 'in_progress'})
-        return pickings_todo.action_assign()
-
-    def cancel_picking(self):
-        self.mapped('picking_ids').action_cancel()
-        return self.write({'state': 'cancel'})
-
-    def print_picking(self):
-        pickings = self.mapped('picking_ids')
-        if not pickings:
-            raise UserError(_('Nothing to print.'))
+        self.state = 'in_progress'
+        return True
+
+    def action_cancel(self):
+        self.ensure_one()
+        self.state = 'cancel'
```

Comment:
remove batch_id from picking? or modify allowed_picking_ids in order to allow picking in canceled batch